src/java.base/share/classes/java/security/Security.java

```diff
@@ -234,13 +234,25 @@ private static void initialize() {
         // FIPS support depends on the contents of java.security so
         // ensure it has loaded first
         if (loadedProps) {
-            boolean shouldEnable = Boolean.valueOf(System.getProperty("com.redhat.fips", "true"));
+            boolean shouldEnable;
+            String sysProp = System.getProperty("com.redhat.fips");
             boolean useProv = Boolean.valueOf(props.getProperty("security.useFIPSProviders"));
             if (sdebug != null ) {
-                sdebug.println("com.redhat.fips=" + shouldEnable);
+                sdebug.println("com.redhat.fips=" + (sysProp == null ? "<unset>" : sysProp));
                 sdebug.println("security.useFIPSProviders=" + useProv);
             }
-            if (shouldEnable && useProv) {
+            if (sysProp == null) {
+                shouldEnable = useProv;
+                if (sdebug != null) {
+                    sdebug.println("com.redhat.fips unset, using security.useFIPSProviders");
+                }
+            } else {
+                shouldEnable = Boolean.valueOf(sysProp);
+                if (sdebug != null) {
+                    sdebug.println("com.redhat.fips set, using its value " + shouldEnable);
+                }
+            }
+            if (shouldEnable) {
```

Comment:
Just for the record, my original idea was to unify this logic as follows, that's why I was asking to also make `-Djava.security.disableSystemPropertiesFile=false` override `java.security`'s content ([in this private comment](https://bugzilla.redhat.com/show_bug.cgi?id=2090378#c11)):

```java
    private static enum SecurityPropertyMeans {
        THE_SAME_AS_THE_SYSTEM_PROPERTY, THE_OPPOSITE_OF_THE_SYSTEM_PROPERTY,
    }
    private static boolean processRedHatSetting(String sysProp, String secProp, SecurityPropertyMeans semantic) {
        final String sysPropValue = System.getProperty(sysProp);
        // When the property is absent in java.security, consider it false,
        // but a cleanly installed java.security could configure it as true
        final String secPropValue = props.getProperty(secProp, "false");
        if (sdebug != null) {
            sdebug.println(secProp + "=" + secPropValue);
            sdebug.println(sysProp + "=" + (sysPropValue == null ? "<unset>" : sysPropValue));
        }
        if (sysPropValue != null) {
            // System properties take precedence, this way, the user can put
            // general settings in java.security for most of the applications,
            // and have a smaller group of applications that override those
            // settings through system properties (e.g. in the command line)
            return Boolean.parseBoolean(sysPropValue);
        } else {
            if (semantic == SecurityPropertyMeans.THE_OPPOSITE_OF_THE_SYSTEM_PROPERTY) {
                return !Boolean.parseBoolean(secPropValue);
            } else {
                return Boolean.parseBoolean(secPropValue);
            }
        }
    }
    // [...]

    // Usages
    boolean configureSysProps = !processRedHatSetting(
            "java.security.disableSystemPropertiesFile",
            "security.useSystemPropertiesFile",
            SecurityPropertyMeans.THE_OPPOSITE_OF_THE_SYSTEM_PROPERTY
    );

    boolean configureFIPS = processRedHatSetting(
            "com.redhat.fips",
            "security.useFipsProviders",
            SecurityPropertyMeans.THE_SAME_AS_THE_SYSTEM_PROPERTY
    );
```

Reply:
> Also, unlike with system security properties, whether FIPS mode support is actually enabled depends on whether the system is in FIPS mode, so saying it is enabled before that is checked could be misleading.

What is being claimed as 'enabled' is the FIPS-mode **\*alignment\*** (with the system's global settings).

Except for that (and in what I understand for 'readability'), I agree with all you say. So, in line with this, given we are decided to change the properties' names, types and semantics:
* Isn't the safest bet in retaining backwards compatibility to just revert 2028344a5c354ac14b8c6ea4ca5a71622cd5e524 and set `security.useSystemPropertiesFile=false` in the portable builds' `java.security`?
* Even when I proposed the `security.useFipsProviders` name, @martinuy made me note that other `java.security` properties typically use the same name as the corresponding system ones (this reduces confusion and makes easier to remember them)
  * Maybe `com.redhat.fips` isn't the best name, so I lean towards thinking of a good name for both the security and system properties in the future
  * But, should we introduce a security property for such a short lifetime? (and should we invest time in documenting it, when we could instead just revert the documentation changes?)
* As I understand from the changes, crypto-policies unalignment now enforces FIPS unalignment (as prior to 2028344a5c354ac14b8c6ea4ca5a71622cd5e524), this means that the use case for this new security property is no longer backed by a customer request (since customers recovered what they were previously using to globally disable FIPS alignment)

This is just my opinion.

Reply:
I've reverted the property addition. The patch now just reverts to the old behaviour, with better debug tracing.